### PR TITLE
[Dot] [MFMA] [FMA] Update Dot implementation to support upstream tests

### PR DESCRIPTION
**Pull request by binarman**
This PR adds:
- support of FP16 outputs for MFMA dot operations
- fallback to FMA implementation in case MFMA can not handle input sizes

### CODE REVIEW

**Comment by zhanglx13** [comment]
@micmelesse Can you try this PR to see if it resolve the issue in test_core?

**Comment by binarman** [comment]
Yes, It is fixing the issue

File: python/test/unit/language/test_core_amd.py [hunk 1]
```diff
@@ -1199,15 +1199,17 @@ def kernel(X, stride_xm, stride_xn,
 
 
 # MFMA Test Dot tests
-@pytest.mark.parametrize("M, N, K, num_warps, col_a, col_b, epilogue, allow_tf32, dtype",
-                         [(*shape, 2, False, False, epilogue, allow_tf32, dtype)
+@pytest.mark.parametrize("M, N, K, num_warps, col_a, col_b, epilogue, allow_tf32, in_dtype, out_dtype",
+                         [(*shape, 2, False, False, epilogue, allow_tf32, in_dtype, out_dtype)
                           for shape in [(64, 64, 64), (32, 32, 32)]
```
Comment (micmelesse) [inline python/test/unit/language/test_core_amd.py:1204]: Why are we not testing (16,16,16)? It is in upstream and when I try it I get a segfault.

**Review by micmelesse — DISMISSED** [review]

File: python/test/unit/language/test_core_amd.py (hunk 1 above)
Comment (binarman) [reply python/test/unit/language/test_core_amd.py:1204]: If you are running them on MI100/200 GPU, it tries to use MFMA instructions with minimal M/N size of 32.

File: python/triton/language/semantic.py [hunk 2]
```diff
@@ -1278,11 +1289,25 @@ def dot(lhs: tl.tensor,
...
+        if lhs.type.scalar.is_int():
+            ret_dot_scalar_ty = tl.int32
+            _0 = builder.create_splat(builder.get_int32(0), [M, N])
+        else:
+            ret_dot_scalar_ty = tl.float32
+            _0 = builder.create_splat(builder.get_fp32(0), [M, N])
+        ret_ty = tl.block_type(ret_dot_scalar_ty, [M, N])
+        ret = tl.tensor(builder.create_dot(lhs.handle, rhs.handle, _0, allow_tf32),
+                        ret_ty)
+        return cast(ret, ret_scalar_ty, builder)
```
Comment (binarman) [inline python/triton/language/semantic.py:1310]: This part is related to support of FP16 output

File: lib/Analysis/Utility.cpp [hunk 3]
```diff
@@ -149,9 +149,29 @@ bool supportMMA(triton::DotOp op, int version) {
...
+  auto bShape = bTy.getShape();
+
+  assert(aShape[1] == bShape[0]);
+  if (!supportMFMAGranularity(aShape[0]) ||
+      !supportMFMAGranularity(aShape[1]) ||
+      !supportMFMAGranularity(bShape[1]))
+    return false;
+
+  auto aElemTy = aTy.getElementType();
+  auto bElemTy = bTy.getElementType();
```
Comment (binarman) [inline lib/Analysis/Utility.cpp:174]: This part enables fallback to FMA implementation for small matrix sizes in C++ part of compiler

File: python/triton/language/semantic.py [hunk 4]
```diff
@@ -1238,9 +1238,20 @@ def gpu_has_mfma() -> bool:
...
+            return True
+    return False
+
 def mfma_supported(M, N, K, allow_tf32, ret_scalar_ty) -> bool:
     if not gpu_has_mfma():
         return False
+    if not mfma_supported_granularity(M) or \
+       not mfma_supported_granularity(N) or \
+       not mfma_supported_granularity(K):
+        return False
```
Comment (binarman) [inline python/triton/language/semantic.py:1254]: This part enables fallback to FMA implementation for small matrix sizes in python part of compiler

File: python/test/unit/language/test_core_amd.py (hunk 1 above)
Comment (binarman) [reply python/test/unit/language/test_core_amd.py]: @micmelesse 
Upd. I've added workaround with FMA instructions, so 16x16x16 tests work

File: lib/Analysis/Utility.cpp [hunk 5]
```diff
@@ -149,13 +149,34 @@ bool supportMMA(triton::DotOp op, int version) {
 }
 
 #ifdef USE_ROCM
+static bool supportMFMAGranularity(int m, int n, int k) {
+  // these limitations are dtype dependent, in future we may relax them
+  int granularityMN = 32;
```
Comment (jayfurmanek) [inline lib/Analysis/Utility.cpp:154]: Should we define these as constants somewhere so we can change easily if needed?

Comment (binarman) [reply lib/Analysis/Utility.cpp:154]: Right, this should be a constant.

About place: I actually can not think of better place for this constants.

P.s. If you are worried about duplication in C++ and python code: I want to refactor this and remove python part eventually, so there will be only one place with these constants.

**Review by jayfurmanek — APPROVED** [review]

**Pull request merged**